### PR TITLE
feat: add NetRC support to the provider installer

### LINKED ISSUE #29349: Allow private provider registries to self-host provider packages/checksums/signatures under the same credentials as the registry itself
I have a remote Terraform registry that experiencing 401 error on the provider download (download_url, shasums_url and shasums_signature_url).
It appears that only for those 3 downloads, the client does not add the authorization header. 

I have one domain for the registry API, the modules and the providers files.
All the other registry API calls are executed with the relevant domain authorization token, as expected.
The modules downloads are executed with the creds in the _.netrc_ file, as expected.
 But the download of the _SHA256SUMS_, _.sig_ and _provider.zip_ files are executed with no authorization header at all.

The host of the registry and the download links are the same.  
(The token for the domain is updated in the _credentials.tfrc.json_ and the creds are in the _.netrc_ file)

I am working with a reverse-proxy between the client to the registry so I have a record of all the request&responses with their headers. 
 
main.tf file:
```
terraform {
 required_providers {
   aws = {
     source = "terraform-remote.mydomain.com/hashicorp/aws"
   }
 }
}
 
module "vpc" {
 source  = "terraform-remote.mydomain.com/terraform-aws-modules/vpc/aws"
 version = "2.21.0"
}
```
[ClientOutTrace.txt](https://github.com/hashicorp/terraform/files/6969318/ClientOutTrace.txt)

Reverse proxy record + the v1/providers/hashicorp/aws/3.53.0/download/darwin/amd64 response: 
ReverseProxyProviderDownloadResponse

Registry call headers (200)
API request headers

The 401 download request headers:
providers request headers

**Terraform version: 1.0.4**

### PR DESCRIPTION
**Pull request by aaklilu**
An [interim solution](https://github.com/hashicorp/terraform/issues/29349#issuecomment-899881338) to lack of credential while downloading provider archives.

Fixes #29349 

## Target Release

1.10.x

### ENHANCEMENTS

 

-  Provider installation now support NetRC and utilises the credentials configured in `.netrc` file. Similar to module downloads.

### CODE REVIEW

**Comment by crw** [comment]
Thanks for this submission!

**Comment by crw** [comment]
Hi @aaklilu, we discussed this in triage today. Generally speaking there is agreement that this is probably the correct approach. The main blocker to doing a full review is maintainer availability. My best guess is that we'll check back in on this PR mid-November to consider a review. I just wanted to set expectations up-front as to timelines. Thanks again for this submission.

**Comment by aaklilu** [comment]
> Hi @aaklilu, we discussed this in triage today. Generally speaking there is agreement that this is probably the correct approach. The main blocker to doing a full review is maintainer availability. My best guess is that we'll check back in on this PR mid-November to consider a review. I just wanted to set expectations up-front as to timelines. Thanks again for this submission.

Thank you @crw for the update.

**Comment by crw** [comment]
Hi @aaklilu, I have been off for the past three weeks so apologies for this late update. As this change would not make a difference for the Hashicorp Terraform Registry (public or private via Enterprise), this change is unlikely to be prioritized in the near future. If you think this is incorrect, especially if you think the Hashicorp enterprise registry could benefit from this change, please let me know as that could change the decision-making process. Thanks very much!

**Comment by aaklilu** [comment]
> Hi @aaklilu, I have been off for the past three weeks so apologies for this late update. As this change would not make a difference for the Hashicorp Terraform Registry (public or private via Enterprise), this change is unlikely to be prioritized in the near future. If you think this is incorrect, especially if you think the Hashicorp enterprise registry could benefit from this change, please let me know as that could change the decision-making process. Thanks very much!

It's fine, thanks for the update @crw

**Comment by liamcervante** [comment]
Hi @aaklilu, I'm happy to review and merge this. The change actually looks fine to me, but before I merge it could you take a look at https://github.com/hashicorp/terraform/pull/36022 and let me know if the documentation update there is actually understandable for someone using this externally? Thanks!

**Review by liamcervante — APPROVED** [review]

**Pull request merged**
